pallets/attesters/src/lib.rs

```diff
@@ -0,0 +1,864 @@
+
+            Ok(())
+        }
+
+        #[pallet::weight(10_000)]
+        pub fn submit_attestation(
+            origin: OriginFor<T>,
+            message: Vec<u8>,
+            signature: Vec<u8>,
+            target: [u8; 4],
+            attestation_for: AttestationFor,
+        ) -> DispatchResult {
+            let account_id = ensure_signed(origin)?;
+
+            // Lookup the attester in the storage
+            let attester = Attesters::<T>::get(&account_id).ok_or(Error::<T>::NotRegistered)?;
+
+            // Check if active set
+            ensure!(
+                ActiveSet::<T>::get().contains(&account_id),
+                Error::<T>::NotActiveSet
+            );
+
+            // Check if the attester is part of the current committee
+            ensure!(
+                CurrentCommittee::<T>::get().contains(&account_id),
+                Error::<T>::NotInCurrentCommittee
+            );
+
+            let is_verified = attester
+                .verify_attestation_signature(ECDSA_ATTESTER_KEY_TYPE_ID, &message, &signature)
+                .map_err(|_| Error::<T>::InvalidSignature)?;
+
+            // todo: slash the attester if the signature is invalid
+            ensure!(is_verified, Error::<T>::AttestationSignatureInvalid);
+
+            match attestation_for {
+                AttestationFor::SFX => {
+                    let sfx_hash: T::Hash = Decode::decode(&mut &message[..])
+                        .map_err(|_| Error::<T>::InvalidMessage)?;
+                    let mut sfx_attestation = Attestations::<T>::get(sfx_hash).unwrap_or_default();
+
+                    // Check if the attester has already signed the attestation
+                    ensure!(
+                        !sfx_attestation
+                            .signature
+                            .iter()
+                            .any(|(attester, _)| attester == &account_id),
+                        Error::<T>::AttestationDoubleSignAttempt
+                    );
+
+                    sfx_attestation
+                        .signature
+                        .push((account_id.clone(), signature));
+
+                    let required_signatures = (T::ActiveSetSize::get() * 2 / 3) as usize;
+                    let status = if sfx_attestation.signature.len() < required_signatures {
+                        AttestationStatus::Pending
+                    } else {
+                        AttestationStatus::Approved
+                    };
+
+                    sfx_attestation.for_ = attestation_for;
+                    sfx_attestation.status = status;
+
+                    Attestations::<T>::insert(sfx_hash, sfx_attestation.clone());
+
+                    let mut target_batches = Batches::<T>::get(target).unwrap_or_default();
+                    let current_block = frame_system::Pallet::<T>::block_number();
+
+                    if target_batches.is_empty() {
```

Comment:
I think this batching logic can be made significantly more efficient by reducing the number of signature checks. IMO the elements that are being signed are queued here, once the queue is full, we throw the elements(sfx, etc) in a merkle tree and let the attestors only sign the root. This way we only do one signature check per attestor and batch